text/src/shaper.rs

```diff
@@ -310,37 +310,54 @@ impl From<TextDirection> for rustybuzz::Direction {
 
 #[derive(Clone)]
 struct FallBackFaceHelper<'a> {
-  ids: &'a [ID],
+  ids: Vec<ID>,
   font_db: &'a RefCell<FontDB>,
   face_idx: usize,
 }
 
 impl<'a> FallBackFaceHelper<'a> {
-  fn new(ids: &'a [ID], font_db: &'a RefCell<FontDB>) -> Self { Self { ids, font_db, face_idx: 0 } }
+  fn new(ids: &'a [ID], font_db: &'a RefCell<FontDB>) -> Self {
+    let mut ids = ids.to_vec();
+    let set: ahash::HashSet<ID> = ahash::HashSet::from_iter(ids.iter().cloned());
+
+    {
+      let font_db = font_db.borrow();
+      let default_ids = font_db.default_fonts();
+      for id in default_ids.iter() {
+        if set.contains(id) {
+          continue;
+        }
+        ids.push(*id);
+      }
+    }
+
+    Self { ids, font_db, face_idx: 0 }
+  }
 
   fn next_fallback_face(&mut self, text: &str) -> Option<Face> {
     let font_db = self.font_db.borrow();
     loop {
-      if self.face_idx > self.ids.len() {
+      if self.face_idx >= self.ids.len() {
         return None;
       }
-      let face_idx = self.face_idx;
       self.face_idx += 1;
-      if face_idx == self.ids.len() {
-        return font_db.try_get_face_data(font_db.default_font()).cloned();
-      }
 
       let face = self
         .ids
-        .get(face_idx)
+        .get(self.face_idx - 1)
```

Comment:
Some suggestion：line 343 move to line 350, can use self.face_idx instead of self.face_idx - 1 at line 347. Is it better?